### PR TITLE
[key] unpack pointers when using key.Infer()

### PR DESCRIPTION
**Pull request by lizthegrey**

### CODE REVIEW

File: api/key/key.go [hunk 1]
```diff
@@ -117,9 +117,48 @@ func Infer(k string, value interface{}) core.KeyValue {
 		return Float64(k, v)
 	case string:
 		return String(k, v)
+	case *bool:
```
Comment (jmacd) [inline api/key/key.go:120]: At the top of the function, consider adding a `if value == nil { return String(k, "nil") }`.

**Comment by vmihailenco** [comment]
>span.SetAttribute("id", &server.id) // avoids an allocation by passing a pointer

Passing small values (up to 24-32 bytes) by pointer is actually slower and can lead to an allocation. Using reflect will make this function 5-6 times slower...

**Comment by vmihailenco** [comment]
>avoids an allocation by passing a pointer

Oh, you mean avoids `interface` allocation. But that is kinda strange point - passing `*int` using `interface{}` instead of passing `int`.

**Comment by jmacd** [comment]
@vmihailenco There is already a form that allows passing `int`, it's `span.SetAttributes(key.Int(...), ...)`, and there are two concerns about this form. (1) it's not as usable as the proposal here, (2) it has to allocate a slice of `[]core.KeyValue` to pass a single argument.

Passing a pointer is an option that an experienced user might use, to avoid a memory allocation, but only in cases where they know that getting the pointer won't cause an allocation.

There's more history here: https://github.com/open-telemetry/opentelemetry-go/issues/302
I was on the losing side of that debate. I think observability libraries should **go out of their way** to allocate as little as possible.

**Comment by jmacd** [comment]
I studied this a bit more, and the Go compiler is able to deduce that the argument to `Infer()` does not escape, which means pointers aren't helping and performance is fine where it stands. We could close this PR, sorry.

As for reflection, it's still quite a useful thing and `Infer()` won't work in a lot of cases without it. I'm surprised you think it'll be that much slower. (I also don't think CPU cost matters as much as memory usage.)

**Comment by jmacd** [comment]
(Hang on, I see a flaw in my own reasoning. I suspect the new `SetAttribute()` interface method will interfere with the escape analysis and there will be an allocation. 

@lizthegrey want to add some benchmarks? These will test the existing `SetAttributes()`:

```
type benchSpan struct {
	trace.NoopSpan

	kvs []core.KeyValue
}

func (s *benchSpan) SetAttributes(kvs ...core.KeyValue) {
	s.kvs = kvs
}

func BenchmarkNoKeyInference(b *testing.B) {
	var sp trace.Span = &benchSpan{}

	b.ReportAllocs()
	b.ResetTimer()

	for i := 0; i < b.N; i++ {
		sp.SetAttributes(key.Int("Attr", 1))
	}
}

func BenchmarkWithKeyInference(b *testing.B) {
	var sp trace.Span = &benchSpan{}

	b.ReportAllocs()
	b.ResetTimer()

	for i := 0; i < b.N; i++ {
		sp.SetAttributes(key.Infer("Attr", 1))
	}
}
```

I measure:

```
BenchmarkNoKeyInference-8     	23790880	        47.0 ns/op	      48 B/op	       1 allocs/op
BenchmarkWithKeyInference-8   	23449888	        48.8 ns/op	      48 B/op	       1 allocs/op
```

I'm interested in running a comparable benchmark with the new `SetAttribute()`. Note that the Infer() does not cause an allocation above, but I suspect it will when passed through `SetAttribute()`.

**Comment by lizthegrey** [comment]
Closing this PR then.

@jmacd I'm rubbish at benchmarking so would prefer you do it given you have a pretty clear idea of what you want to test, etc.

**Pull request closed without merging**